### PR TITLE
fpga_crihook: check if AFU is already programmed

### PR DESCRIPTION
**Pull request by bart0sh**
Don't program device if it's already programmed with requested AFU.

### CODE REVIEW

File: cmd/fpga_crihook/main.go [hunk 1]
```diff
@@ -44,54 +48,56 @@ func decodeJSONStream(reader io.Reader) (map[string]interface{}, error) {
 }
 
 type hookEnv struct {
-	bitStreamDir string
-	config       string
-	execer       utilsexec.Interface
+	bitStreamDir  string
+	config        string
+	execer        utilsexec.Interface
+	aFUIDTemplate string
```
Comment (rojkov) [inline cmd/fpga_crihook/main.go]: formatting...

File: cmd/fpga_crihook/main.go [hunk 2]
```diff
@@ -34,6 +36,8 @@ const (
 	fpgaRegionEnv          = "FPGA_REGION"
 	fpgaAfuEnv             = "FPGA_AFU"
 	fpgaBitStreamExt       = ".gbs"
+	fpgaDevRegexp          = `\/dev\/intel-fpga-port.(\d)$`
+	aFuIDTemplate          = "/sys/class/fpga/intel-fpga-dev.%s/intel-fpga-port.0/afu_id"
```
Comment (rojkov) [inline cmd/fpga_crihook/main.go]: why capital F in `aFuIDTemplate`?

File: cmd/fpga_crihook/main.go [hunk 3]
```diff
@@ -44,54 +48,56 @@ func decodeJSONStream(reader io.Reader) (map[string]interface{}, error) {
...
+	aFUIDTemplate string
 }
 
-func newHookEnv(bitStreamDir string, config string, execer utilsexec.Interface) *hookEnv {
+func newHookEnv(bitStreamDir string, config string, execer utilsexec.Interface, aFUIDTemplate string) *hookEnv {
 	return &hookEnv{
 		bitStreamDir,
 		config,
 		execer,
+		aFUIDTemplate,
```
Comment (rojkov) [inline cmd/fpga_crihook/main.go]: capital FU?

File: cmd/fpga_crihook/main.go [hunk 4]
```diff
@@ -164,10 +189,29 @@ func (he *hookEnv) programBitStream(fpgaRegion string, fpgaAfu string, fpgaBitSt
 	return nil
 }
 
+func (he *hookEnv) getProgrammedAfu(deviceNum string) (string, error) {
+	aFuIDPath := fmt.Sprintf(he.aFUIDTemplate, deviceNum)
+	data, err := ioutil.ReadFile(aFuIDPath)
```
Comment (rojkov) [inline cmd/fpga_crihook/main.go]: F and FU

File: cmd/fpga_crihook/main_test.go [hunk 5]
```diff
@@ -26,18 +26,21 @@ import (
...
 	tcases := []struct {
-		stdinJSON      string
-		configJSON     string
-		expectedErr    bool
-		expectedRegion string
-		expectedAFU    string
+		stdinJSON         string
+		configJSON        string
+		afuIDPath         string
+		expectedErr       bool
```
Comment (rojkov) [inline cmd/fpga_crihook/main_test.go]: formatting...

File: cmd/fpga_crihook/main.go [hunk 6]
```diff
@@ -164,10 +189,29 @@ func (he *hookEnv) programBitStream(fpgaRegion string, fpgaAfu string, fpgaBitSt
...
+	if err != nil {
+		return "", err
+	}
+	return strings.TrimSpace(string(data)), nil
+}
+
 func (he *hookEnv) process(reader io.Reader) error {
-	fpgaRegion, fpgaAfu, err := he.getFPGAParams(reader)
+	fpgaRegion, fpgaAfu, deviceNum, err := he.getFPGAParams(reader)
```
Comment (rojkov) [inline cmd/fpga_crihook/main.go]: The `fpga` prefix in `fpgaAfu` is redundant. It's already clear that we deal with FPGA here.

File: cmd/fpga_crihook/main_test.go [hunk 7]
```diff
@@ -262,29 +290,50 @@ func TestProcess(t *testing.T) {
...
+		{
+			stdinJSON:     "stdin-correct.json",
+			configJSON:    "config-non-existing-bitstream.json",
+			aFuIDTemplate: "testdata/sys/class/fpga/intel-fpga-dev.%s/intel-fpga-port.0/afu_id_not_programmed_yet",
+			expectedErr:   true,
+		},
+		{
+			stdinJSON:     "stdin-correct.json",
+			configJSON:    "config-correct.json",
+			aFuIDTemplate: "testdata/sys/class/fpga/intel-fpga-dev.%s/intel-fpga-port.0/afu_id_not_programmed_yet",
```
Comment (rojkov) [inline cmd/fpga_crihook/main_test.go]: F

File: cmd/fpga_crihook/main.go (hunk 2 above)
Comment (rojkov) [reply cmd/fpga_crihook/main.go]: Let's assume there are two FPGA devices. Then there will be dev nodes
```
/dev/intel-fpga-port.0
/dev/intel-fpga-port.1
```
In this case the path to `afu_id` will be `/sys/class/fpga/intel-fpga-dev.1/intel-fpga-port.1/afu_id`, not `/sys/class/fpga/intel-fpga-dev.1/intel-fpga-port.0/afu_id`

File: cmd/fpga_crihook/main.go [hunk 8]
```diff
@@ -44,54 +48,56 @@ func decodeJSONStream(reader io.Reader) (map[string]interface{}, error) {
...
+		aFUIDTemplate,
 	}
 }
 
 func canonize(uuid string) string {
 	return strings.ToLower(strings.Replace(uuid, "-", "", -1))
 }
 
-func (he *hookEnv) getFPGAParams(reader io.Reader) (string, string, error) {
+func (he *hookEnv) getFPGAParams(reader io.Reader) (string, string, string, error) {
```
Comment (rojkov) [inline cmd/fpga_crihook/main.go]: I guess now these 3 strings in the function's return would be better joined into a struct like
```
type fpgaParams {
    region string
    afu string
    devNum string
}
```

File: cmd/fpga_crihook/main.go (hunk 1 above)
Comment (bart0sh) [reply cmd/fpga_crihook/main.go]: what's wrong with it? gofmt doesn't complain

Comment (rojkov) [reply cmd/fpga_crihook/main.go]: weird... I clearly see in my browser that the type names in the struct are not aligned, yet `git diff` shows no problems with that. Must be a browser bug. Disregard then.

File: cmd/fpga_crihook/main.go (hunk 8 above)
Comment (bart0sh) [reply cmd/fpga_crihook/main.go]: this caused 208 lines of changes so far. If I'd know that I'd not even start. It doesn't worth it in my opinion.

File: cmd/fpga_crihook/main.go [hunk 9]
```diff
@@ -102,85 +114,123 @@ func (he *hookEnv) getFPGAParams(reader io.Reader) (string, string, error) {
...
 	output, err := he.execer.Command("fpgaconf", fpgaBitStreamPath).CombinedOutput()
 	if err != nil {
-		return fmt.Errorf("failed to program AFU %s to region %s: error: %v, output: %s", fpgaAfu, fpgaRegion, err, string(output))
+		return fmt.Errorf("failed to program AFU %s to region %s: error: %v, output: %s", params.afu, params.region, err, string(output))
 	}
 	return nil
 }
 
+func (he *hookEnv) getProgrammedAfu(deviceNum string) (string, error) {
+	afuIDPath := fmt.Sprintf(he.afuIDTemplate, deviceNum, deviceNum)
```
Comment (rojkov) [inline cmd/fpga_crihook/main.go]: I'd leave a message for future us that this code supports only one AFU per device to give us a clue when the hook starts misbehaving with more advanced hardware.

Comment (bart0sh) [reply cmd/fpga_crihook/main.go]: do you mean adding a comment or print a message?

Comment (rojkov) [reply cmd/fpga_crihook/main.go]: a comment would suffice I think

Comment (bart0sh) [reply cmd/fpga_crihook/main.go]: done

**Pull request merged**